adaptdl/adaptdl/torch/adascale.py

```diff
@@ -77,96 +81,72 @@ class AdaScale(object):
 
     .. _AdaScale: https://proceedings.icml.cc/static/paper_files/icml/2020/4682-Supplemental.pdf
     """  # noqa: E501
-    def __init__(self, optimizer, scale=None,
-                 num_replicas=None, patch_optimizer=False):
+    def __init__(self, adp, optimizer, num_replicas=None,
+                 accum_scale=None, patch_optimizer=False):
+        self._adp = adp
         self._optimizer = optimizer
-        self._optimizer_step = optimizer.step
-        self._local_sqr = None
+        self._optimizer_step = None
+        self._optimizer_zero_grad = None
         self._num_replicas = (num_replicas if num_replicas is not None
                               else torch.distributed.get_world_size())
+        self._accum_scale = accum_scale or self._num_replicas
         self._prev_grads = None
-        self._made_step = False
-        self._accumulation_steps = 1
-        self._current_accumulation_step = 0
 
         self._optimizer.state.setdefault("adascale", {
-            "replicas": 0.0,
-
+            "progress": 0.0,
+            "prev_scale": 0.0,
             # Averages of n and v
-            "norm_avg": np.ones(len(optimizer.param_groups)),
+            "sqr_avg": np.ones(len(optimizer.param_groups)),
             "var_avg": np.zeros(len(optimizer.param_groups)),
+            # Whether estimates are biased (using differenced estimator).
+            "biased": False,
         })
 
-        self.set_scale(self._num_replicas if scale is None else scale)
-
         for idx, param_group in enumerate(self._optimizer.param_groups):
             for param in param_group["params"]:
                 param.register_hook(
-                    functools.partial(self._backward_hook, idx))
+                    functools.partial(self._backward_hook, idx, param))
+        self._callback_queued = False
 
         if patch_optimizer:
             self.patch_optimizer()
 
         self._smoothing = 0.999
 
+        self.raw_grad_sqr = np.ones(len(optimizer.param_groups))
+        self.raw_grad_var = np.zeros(len(optimizer.param_groups))
+
+        self._reset_accumulation()
+
+    def _reset_accumulation(self):
+        if self._optimizer_zero_grad is None:
```

Comment:
I'm not a huge fan of this change to be honest. I don't think its a good idea to fully automate zeroing of gradients and to force zeroing during the `step` call.